### PR TITLE
Remove type arguments from error types

### PR DESCRIPTION
**Pull request by edsko**
NOTE:  "Remove `tip` argument to ChainSyncClient" is #1238.

This commit paves the way for consensus error policies. It removes type parameters from exception types, so that we cannot introduce bugs where we miss certain exceptions because due to (type) matching on exceptions of the wrong type (e.g., instantiated to header instead of block). 

It also removes `onSlot` from `BlockchainTime`. This was used only in tests, where it is easily replaced.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs [hunk 1]
```diff
@@ -106,38 +91,6 @@ blockUntilSlot btime slot = do
...
-                   else Nothing
-
-data OnSlotException =
-    -- | An action was scheduled via 'onSlot' for a slot in the past.
-    -- First slot is requested, second slot is current as of raising.
-    OnSlotTooLate SlotNo SlotNo
-  deriving (Eq, Show)
-
-instance Exception OnSlotException
-
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: @nfrisby I remember you added this. What do you think about this change?

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs [hunk 2]
```diff
@@ -841,93 +841,151 @@ continueWithState !s (Stateful f) = checkInvariant (unsafeNoThunks s) $ f s
...
+    case eqT @blk @blk' of
+      Nothing   -> False
+      Just Refl -> (a, b, c, d) == (a', b', c', d')
+  DoesntFit{} == _ = False
+
+  InvalidBlock (a :: Point blk) == InvalidBlock (a' :: Point blk') =
+    case eqT @blk @blk' of
+      Nothing   -> False
+      Just Refl -> a == a'
+  InvalidBlock{} == _ = False
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: :scream:

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: (If the exception is needed, we can re-introduce it in the test.)

File: ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/ChainSyncClient.hs]: I know :(

**Comment by edsko** [comment]
Yes, will do. Once 1238 is merged, will rebase and reintroduce the exception in the consensus tests.

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: Rebased, and moved this to `Test.Util.BlockchainTime` where it is now defined as

```haskell
onSlot :: (HasCallStack, IOLike m) => BlockchainTime m -> SlotNo -> m () -> m ()
onSlot btime slot k = do
    startingSlot <- atomically $ getCurrentSlot btime
    when (startingSlot >= slot) $
      throwM $ OnSlotTooLate slot startingSlot
    onSlotChange btime $ \slot' ->
      when (slot == slot') k

data OnSlotException =
    -- | An action was scheduled via 'onSlot' for a slot in the past.
    -- First slot is requested, second slot is current as of raising.
    OnSlotTooLate SlotNo SlotNo
  deriving (Eq, Show)

instance Exception OnSlotException
```

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs]: Moving it is no problem. Thanks for the notice.

**Pull request merged**